ethcore/src/engines/clique/tests.rs

```diff
@@ -0,0 +1,768 @@
+		let state = self.clique.state(header).unwrap();
+		if state.is_inturn(block_num, signer) {
+			DIFF_INTURN
+		} else {
+			DIFF_NOTURN
+		}
+	}
+
+	// Note, this will read the cache and `will` not work with more than 128 blocks
+	fn get_state_at_block(&self, hash: &H256) -> CliqueBlockState {
+		self.clique.block_state_by_hash.write()
+			.get_mut(hash)
+			.expect("CliqueBlockState not found tested failed")
+			.clone()
+	}
+
+	// Get signers after a certain state
+	// This is generally used to fetch the state after a test has been executed and checked against
+	// the intial list of signers provided in the test
+	fn clique_signers(&self, hash: &H256) -> Vec<Address> {
+		self.get_state_at_block(hash).signers().to_vec()
+	}
+
+	// Fetches all addresses at current `block` and converts them back to `tags (char)` and sorts them
+	// Addresses are supposed sorted based on address but these tests are using `tags` just for simplicity
+	// and the order is not important!
+	fn tags_from_vec(&self, addr: &Vec<Address>) -> Vec<char> {
+		let mut tags: Vec<char> = addr.iter().filter_map(|addr| {
+			for (t, kp) in self.signers.iter() {
+				if addr == &kp.address() {
+					return Some(*t)
+				}
+			}
+			None
+		}).collect();
+		tags.sort();
+		tags
+	}
+
+	fn new_block_and_import(
+		&self,
+		block_type: CliqueBlockType,
+		last_header: &Header,
+		beneficary: Option<Address>,
+		signer: char,
+		) -> Result<OpenBlock, Error> {
+
+		let mut extra_data = vec![0; VANITY_LENGTH];
+		let mut seal = util::null_seal();
+		let last_hash = last_header.hash();
+
+		match block_type {
+			CliqueBlockType::Checkpoint => {
+				let signers = self.clique.state(&last_header).unwrap().signers().clone();
+				for signer in signers {
+					extra_data.extend(&*signer);
+				}
+			}
+			CliqueBlockType::Vote(v) => seal = v.as_rlp(),
+			CliqueBlockType::Empty => (),
+		};
+
+		let db = self.db.boxed_clone();
+
+		let mut block = OpenBlock::new(
+			&self.clique,
+			Default::default(),
+			false,
+			db,
+			&last_header.clone(),
+			Arc::new(vec![last_hash]),
+			beneficary.unwrap_or_default(),
+			(3141562.into(), 31415620.into()),
+			extra_data,
+			false,
+			&mut Vec::new().into_iter(),
+		).unwrap();
+
+		{
+			let difficulty = self.get_difficulty(block.header().number(), last_header, &self.signers[&signer].address());
+			let b = block.block_mut();
+			b.header.set_timestamp(last_header.timestamp() + self.clique.period);
+			b.header.set_difficulty(difficulty);
+			b.header.set_seal(seal);
+
+			let sign = ethkey::sign(self.signers[&signer].secret(), &b.header.hash()).unwrap();
+			let mut extra_data = b.header.extra_data().clone();
+			extra_data.extend_from_slice(&*sign);
+			b.header.set_extra_data(extra_data);
+		}
+
+		self.clique.verify_block_basic(block.header())?;
+		self.clique.verify_block_family(block.header(), &last_header)?;
+
+		Ok(block)
+	}
+}
+
+#[test]
```

Comment:
I think we should have a link to the test cases (http://eips.ethereum.org/EIPS/eip-225) just so people know that we didn't make them up ourselves